### PR TITLE
Add indicator dm020

### LINKED ISSUE #5: Write ehrQL code for DM020 indicator
## Denominator

- [x] DM020 rule 1: Reject patients from the specified population whose latest frailty diagnosis is moderate or severe. Pass all remaining patients to the next rule. (NOTE: this seems to be different to DM020 rule 1, see #10)

- [x] DM020 rule 2: Select patients passed to this rule who meet both of the criteria below:

  
  - Have their latest IFCC-HbA1c of 58 mmol/mol or less.
  - Have their latest IFCC-HbA1c reading was recorded in the 12 months leading up to and including the payment period end date.
    
  Pass all remaining patients to the next rule.

- [x] DM020 rule 3: Reject patients passed to this rule who did not have their IFCC-HbA1c recorded during the current service year but did have serum fructosamine recorded during the service year. Pass all remaining patients to the next rule.

- [x] DM020 rule 4: Reject patients passed to this rule who are on maximum tolerated diabetes treatment in the 12 months leading up to and including the payment period end date. Pass all remaining patients to the next rule.

- [x] DM020 rule 5: Reject patients passed to this rule for whom diabetes quality indicator care was unsuitable in the 12 months leading up to and including the payment period end date. Pass all remaining patients to the next rule.

- [x] DM020 rule 6: Reject patients passed to this rule who chose not to receive a blood test in the 12 months leading up to and including the payment period end date. Pass all remaining patients to the next rule.

- [x] DM020 rule 7: Reject patients passed to this rule who chose not to receive diabetes quality indicator care in the 12 months leading up to and including the payment period end date. Pass all remaining patients to the next rule.

- [x] DM020 rule 8: Reject patients passed to this rule who meet either of the criteria below:
  - Latest IFCC-HbA1c recorded in the 12 months leading up to and including the payment period end date was above target levels (IFCC-HbA1c was above 58 mmol/mol), and was followed by two invitations for diabetes monitoring.
  - Received two invitations for diabetes monitoring and had no IFCC-HbA1c recorded during the 12 months leading up to and including the achievement date.
    
  Pass all remaining patients to the next rule.
  _Please note: To support GP practices in enabling them to carry out QOF care reviews after two invitations for care have been coded, clinical system prompts should not remove a patient from the indicator._ 
  
  _Currently, once the second invite code has been added to the patient record, the Business Rules logic will remove the patient from the Denominator and clinical systems may automatically remove the prompt, removing the chance for opportunistic reviews where necessary. If a patient has not responded to the invitations, and has not received the intervention required by a given indicator, then that patient will be removed from the indicator via personalised care adjustments at year end. This is for payment purposes only. However, up until year end it is expected that such patients should still be flagged up to practices as requiring the care specified by the indicator._

- [x] DM020 rule 9: Reject patients passed to this rule whose diabetes diagnosis was in the 9 months leading up to and including the payment period end date. Pass all remaining patients to the next rule.

- [x] DM020 rule 10: Reject patients passed to this rule who registered with the practice in the 9 months leading up to and including the payment period end date. Select the remaining patients.

## Numerator

- [x] DM021 rule 1: Select patients passed to this rule who meet both of the criteria below:

  - Have their latest IFCC-HbA1c of 58 mmol/mol or less.
  - Have their latest IFCC-HbA1c reading was recorded in the 12 months leading up to and including the payment period end date.

Reject the remaining patients.

### PR DESCRIPTION
**Pull request by milanwiedemann**
This PR makes some changes to the setup of the study code (i.e., naming conventions, shared variables):

- **Naming conventions**: I'm aiming to stick to the names used in the source document (QOF Business Rules for Diabetes), e.g., `dm_res_cod` (slightly easier to read) was changed back to  `dmres_cod` (harder to read but consistent with the source document and therefore less thinking needed when translating the rules into code)
- **Shared variables**: the (i) variables for all indicators (Section 3.2.4 in the QOF Business Rules) as well as the (ii) business rules (Section 4.1 in QOF Business Rules) get defined in a shared variable dictionary `analysis/variable_lib_dm.py`.  The variable library is structured into two main parts. Although the code in this file is heavily documented, it is probably still hard to understand without opening the diabetes business rules (either _Diabetes_v46.0.docx_ in our shared drive or from the .zip file available [here](https://digital.nhs.uk/data-and-information/data-collections-and-data-sets/data-collections/quality-and-outcomes-framework-qof/quality-and-outcome-framework-qof-business-rules/qof-business-rules-v46.0-2021-2022-baseline-release)).
  1. First, the function `add_diabetes_mellitus_register_variables()` adds all variables specified in the business rules to a dataset. We might want to add the option to only add the variables needed for a particular indicator at a later stage.
  2. Second, the business rule logic also gets defined in functions, e.g., `get_dm_reg_r1` gets the business rule logic for the diabetes mellitus (`dm`) register (`reg`) rule 1 (`r1`). I'm checking a lot for null values before doing date or other comparisons and I'm not sure this is needed. when writing the business rules it really payed off to use the original (cryptic) variable names from the source document.

To _test_ the business rules I manually created dummy data in various `.csv` files inside the `dummy_data/` subfolder. This process was VERY laborious but payed off! I'm currently expecting dummy patients 2, 3, 5, and 8 to be in the register (dm017). Dummy patients 2 (select rule number 2) and 8 (select rule number 10) are selected for indicator dm020.

---

I followed these steps to add a the new dm020 indicator, this process should be similar for further indicators:

1. Add new variables (e.g., `dmres_dat`: "Date of the most recent diabetes diagnosis resolved code") to the shared variable library `analysis/variable_lib_dm.py` 
2. Translate QOF business rules for dm020 (e.g., Rule 1: "Reject patients from the specified population whose
 latest frailty diagnosis is moderate or severe.")  and add these also to `analysis/variable_lib_dm.py`. As mentioned, some rules are nearly identical between indicators so I added arguments to specify the values that are different. I also provided the option to specify an `index_date` to allow us to run this study monthly once we have a feature in databuilder that allows us to do something similar to `--index_date_range` from cohortextractor.
4. Create dummy data test patients that will help us test if the QOF rules we are implementing (see next step) are selecting/rejecting the correct patients
5. Create a new file for this dataset definition `analysis/dataset_definition_dm020.py` and start adding the new business rules step by step while running `opensafely run run_all -f` each time I introduce a new rule to check that it does exactly what I was expecting

---

This PR makes it very easy to add similar indicators, e.g., dm021 (see #11)

### CODE REVIEW

**Review by inglesp — CHANGES_REQUESTED** [review]
There's a lot to like here (not least the incredibly helpful PR description), and I can see the benefits of having a separate library for the rules, which you then use in the different dataset definitions.

However, the approach you've taken doesn't look quite right -- but it shouldn't be too hard to fix, and none of your work to convert QOF rules into ehrQL is wasted.

The issue is that unclear from the code about where the `dataset` variable comes from and what it represents.  The data flow seems to be:

* in variable_lib_dm.py, create a `Dataset` instance and call this `dataset` in the module namespace
* in dataset_definition_dm020.py, import this thing called `dataset`
* in dataset_definition_dm020.py, pass `dataset` to `variable_lib_dm.add_diabetes_mellitus_register_variables`, and call whatever this function returns `dataset` (although it's actually the same object that `dataset` originally referrred to, so this is unnecessary)
* in dataset_definition_dm020.py, call various functions (eg `variable_lib_dm.get_dm_reg_r1`) without passing `dataset` to these functions, and instead using the thing called `dataset` in the namespace of the `variable_lib_dm` module, which happens to be the same object as the thing called `dataset` in the `dataset_definition_dm020` module

It doesn't actually seem to be a problem for this code, since it apparently works and gives you the answers you expect, but I think it would be a source of confusion (it certainly took me a few minutes of reading to make sure I'd understood exactly what was going on) which risks letting errors creep in if we use this pattern more widely.

One solution would be to instantiate the `Dataset` class in dataset_definition_dm020.py, and always pass a `dataset` parameter to each of your variable_lib_dm.py functions:

```
# dataset_definition_dm020.py

dataset = Dataset()
add_diabetes_mellitus_register_variables(dataset, index_date)  # note we don't change what `dataset` refers to here
dataset.dm_reg_r1 = get_dm_reg_r1(dataset)
```

and

```
# variable_lib_dm.py

add_diabetes_mellitus_register_variables(dataset, index_date):
    ...
    # don't return anything

def get_dm_reg_r1(dataset):
    return (dataset.dmres_dat < dataset.dmlat_dat) | (
        dataset.dmlat_dat.is_not_null() & dataset.dmres_dat.is_null()
    )
```

However, a more idiomatic approach might be to subclass `Dataset` with some helper methods in variable_lib_dm.py, and then instantiate that subclass in the indicator modules, and call the methods to set the variables on the dataset instance:

```
# dm_dataset.py (I prefer this to variable_lib_dm.py)

from databuilder import Dataset

class DmDataset(Dataset):
    def add_register_variables(self, index_date):
        prior_events = get_prior_events(index_date)
        self.reg_dat = practice_registration_as_of(index_date).start_date
        ...

    def add_dm_reg_r1(self):
        self.dm_reg_r1 = (self.dmres_dat < self.dmlat_dat) | (
            self.dmlat_dat.is_not_null() & self.dmres_dat.is_null()
        )
```

and

```
# dataset_definition_dm017.py

from dm_dataset import DmDataset

dataset = DmDataset()
dataset.add_register_variables(index_date)
dataset.add_dm_reg_r1()
...
```

What do you think?  I don't know whether you've encountered Python's class/object system before.  If this is new to you, I can talk you through it next week.

**Comment by milanwiedemann** [comment]
I've addressed two points since your initial review:

1. The approach to use one subclass `class DmDataset(Dataset):`  for both the variables and business rules seems to be the the best one for now (instead of one subclass for the business rules and one for the variables).
    - First, all QOF variables can be added using:

    ```
    dataset = DmDataset()
    dataset.add_diabetes_mellitus_variables(index_date)
    ```

    - Next, business rules can be added to each dataset definition and combined as neede:

    ```
    # DM_REG rule 1:
    dataset.dm_reg_r1 = dataset.get_dm_reg_r1()

    # DM_REG rule 2:
    dataset.dm_reg_r2 = dataset.get_dm_reg_r2()

    dm_reg_select_r2 = dataset.dm_reg_r1 & ~dataset.dm_reg_r2
    ```

2. I decided to name all business rules based on the first time they appear in the business rules document. This is not ideal but seem to be the easiest solution (for now). At this point other options, e.g., (i) writing code for each business rule, or (ii) writing a function that allow the user to generate business rules logic would either result in code duplication or requite a complicate framework that may take considerable time to develop. While I think we should certainly keep the idea of a QOF rule generator in mind, I think we shouldn't optimise something we don't fully understand yet (we dont know all the edge cases of all QOF rules yet).

**Comment by milanwiedemann** [comment]
just spoke to @iaindillingham, I'll add more information (e.g., index_date, cutoff_values) into the DmDataset subclass using something like

```
    def __init__(self, index_date, cut_off_1):
        self.index_date = index_date
        self.cut_off_1 = cut_off_1
        super().__init__()
```

**Comment by milanwiedemann** [comment]
@inglesp following discussion with Dave on slack I'm now using a factory function instead of subclassing Dataset from databuilder. any more suggestions from you or can I merge now?

I will create duplicates of the functions that define identical rules across indicators when I'm adding a new indicator

File: analysis/dm_dataset.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+import codelists
+
+
+# Define a dataset including all variables needed for the diabetes mellitus (dm)
+def make_dm_dataset(
+    index_date: date,
+    ifcchba_cutoff_val: float = None,
+    bpsys_cutoff_val: float = None,
+    bpdia_cutoff_val: float = None,
```
Comment (inglesp) [inline analysis/dm_dataset.py]: Apart from `index_date`, the function parameters are unused.  Can we drop them?

File: analysis/dm_dataset.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    dataset.sevfrail_dat = last_matching_event(
+        prior_events, codelists.sevfrail_cod
+    ).date
+    # Field number: 66
+    # FRAILLAT_DAT: The date of the latest frailty diagnosis up to and
+    # including the achievement date.
+    mildmodsev_frail_cod = combine_codelists(
+        codelists.mildfrail_cod, codelists.modfrail_cod, codelists.sevfrail_cod
+    )
+    dataset.fraillat_dat = last_matching_event(prior_events, mildmodsev_frail_cod).date
```
Comment (inglesp) [inline analysis/dm_dataset.py:157]: Possibly just personal preference, so please ignore if you disagree...  I find this long block of continuous code hard to read, and would prefer it if there was an empty line before each `# Field number:` line.

**Review by inglesp — APPROVED** [review]
Looks good, well done.  I've made a couple of small suggestions, but from a code perspective, I think you should merge it and move on.

I've not checked the implementation of all the rules very carefully.  How do you think we should do this?

File: analysis/dm_dataset.py (hunk 1 above)
Comment (milanwiedemann) [reply analysis/dm_dataset.py]: yes, good point, thx for spotting

File: analysis/dm_dataset.py (hunk 2 above)
Comment (milanwiedemann) [reply analysis/dm_dataset.py:157]: true, I agree and have changed it

**Comment by milanwiedemann** [comment]
> I've not checked the implementation of all the rules very carefully. How do you think we should do this?

I think there are 2 parts to this:
1. checking that the translation of the clinical rules into ehrQL is correct. we probably need another pair of eyes, but I also think that writing tests (see next step) will help to catch any translation errors I made
2. testing that the code does exactly what I want it does (provided step 1 is correct). in #12  I moved the creation of test dummy data patients into an R script so that I can add more comments and was thinking to add a pytest to test the output files for each indicator

**Pull request merged**